CL2-6817 BE Only set registration_completed_at from complete_registration method

jinjagit opened the pull request:
## Checklist

- [ ] Added entry to changelog
  
NO CHANGELOG ENTRY - this PR does not constitute a completed feature or bugfix, but is part of a larger piece of work.

- [ ] Tests
<details>
<summary>More info</summary>

### Unit tests

Did you add relevant unit tests?

### E2E tests

Sometimes it can be more efficient to update E2E tests after CI has run them. If you know which ones to update, go ahead! E2E template cl2-back:

```bash
docker compose run --rm web bin/rails cl2_back:create_tenant[localhost,e2etests_template]
```

</details>

- [x] Prepared branch for code review
<details>
<summary>More info</summary>
Reviewed code to reduce unnecessary back and forth (removal of console.log, comments, ...)? Added comments to clarify code, emphasize what to pay attention to, etc.?
</details>

## Links

- [Jira ticket](https://citizenlab.atlassian.net/browse/CL2-6817)

## What changes are in this PR?

Removes writes to `registration_completed_at` field in User record, and related calls to code that writes the value.

**Aim:** To limit the setting of a value for `registration_completed_at` to only via the `complete_registration` action in `users_controller.rb`, which in turn enables a more robust signup and login flow where the FE can always depend on this attribute to inform which step of the signup or login flow should be used (if any).

**IMPORTANT!** This is a very significant change to mission-critical processes (signup & login). These changes must be thoroughly tested in staging before deciding to merge to master. Such testing should, at least, include the email and FB/Google SSO signup and login flows, but probably should also include those verification methods that can also complete/inform the registration flow.

## How urgent is a code review?

ASAP: FE work to fix a bug + make signup & login flow more robust needs these changes.

alexander-cit committed:
Specify poll_interval to speed up Que integration tests

alexander-cit committed:
Merge pull request #1038 from CitizenLabDotCo/fix-app-job-spec-timeouts

Specify poll_interval to speed up Que integration tests

alexander-cit committed:
Annotate models in engines

alexander-cit committed:
Merge pull request #1044 from CitizenLabDotCo/annotate-models-in-engines

Annotate models in engines

sebastienhoorens committed:
dont validate title lenght of ideas in backend

sebastienhoorens committed:
Merge branch 'master' of github.com:CitizenLabDotCo/citizenlab

sebastienhoorens committed:
dont validate title lenght of ideas in backend

sebastienhoorens committed:
bulk import ideas script that deals with duplicate topics

jinjagit committed:
Remove from SideFxUserService and related calls to removed code

Removes before_create & timestamp_registration + related patch
Removes calls to before_create from v1 & admin_api
users_controller files

jinjagit committed:
Remove 2 calls to removed SideFXUserService.before_create

Removed from omniauth_callback_controller create method &
invites_service save_invtes method

jinjagit committed:
Remove tests and remove a test expectation

Removes tests related to SideFxUserService.before_create, which
no longer exists, and removes expectation that
registration_completed_at will be set when a user is created.

jinjagit committed:
Remove from SideFxInviteService & remove patch file

Removes conditional write to registration_completed_at from
back/app/services/side_fx_invite_service.rb and removes file:
back/engines/free/user_confirmation/app/services/...
user_confirmation/patches/side_fx_invite_service.rb as the only
function of the patch was a conditional write to
registration_completed_at.

jinjagit committed:
Remove prepend of non-existent patch file

jinjagit committed:
Remove test expectation

Removes expectation from test in invites_spec that
registration_completed_at will be set

jinjagit committed:
Replace patch file side_fx_invite_service

Eroneously removed, but should remain as conditionally sets
email_confirmed_at, not registration_completed_at (as
previously thought)

jinjagit committed:
Replace prepend of replaced patch file

Erroneously removed this prepend, when erroneously removed
the related patch file. Now patch file is replaced we also
replace this prepend statment.

jinjagit committed:
Remove complete_registration_without_custom_fields from confirm_user.rb

jinjagit commented:
@kogre A couple of questions about related `LogActivityJob`s and testing...  
 
`'completed_registration'` `LogActivityJob` occurences:  
  
Should we also remove the `'completed_registration'` `LogActivityJob`s from [here](https://github.com/CitizenLabDotCo/citizenlab/blob/c9ced3205b6661b6ebbf9808308ab329050f6cda/back/app/services/side_fx_user_service.rb#L15), [here](https://github.com/CitizenLabDotCo/citizenlab/blob/c9ced3205b6661b6ebbf9808308ab329050f6cda/back/app/services/side_fx_user_service.rb#L29), and [here](https://github.com/CitizenLabDotCo/citizenlab/blob/c9ced3205b6661b6ebbf9808308ab329050f6cda/back/app/services/side_fx_invite_service.rb#L19)? (since, I _think_, we now expect these conditions to never be true)...

And should we add a `'completed_registration'` `LogActivityJob` to the [`complete_registration`](https://github.com/CitizenLabDotCo/citizenlab/blob/c9ced3205b6661b6ebbf9808308ab329050f6cda/back/app/controllers/web_api/v1/users_controller.rb#L143) action in `users_controller.rb`, perhaps in the `if @user.save` block? (I'm kinda surprised we don't do that already)  

Testing:  

Is there any need for additional FE testing to cover the removals of BE responsibility for writing to `registration_completed_at` in certain special cases? Because I don't yet understand our FE testing, I can't answer this.

jinjagit commented:
@kogre I will set the base branch to merge with to: Luuc's [CL2-6817-registration-fb-google-sso-no-reg-questions](https://github.com/CitizenLabDotCo/citizenlab/tree/CL2-6817-registration-fb-google-sso-no-reg-questions) when this code is reviewed and agreed 'ready'. I have not done this now as it will add approx. 100 more files in the 'Files changed' list, which is not super-helpful to the reviewer.

kogre commented on back/app/services/side_fx_user_service.rb:
```diff
@@ -3,10 +3,6 @@
 
-  def before_create(user, current_user)
-    timestamp_registration(user)
-  end
```
I would keep this method and just make it empty, and also keep every place that's calling it.

This is debatable, and if it would just be called from one location I think removing it completely would be fine. But in this case, losing when exactly this side effect should be called from the various places really is core to the business logic (were in the system are we creating a user?) that it could be painful to lose if we need new side effects on before_create of a user.

kogre commented on back/app/controllers/omniauth_callback_controller.rb:
```diff
@@ -66,7 +66,6 @@ def auth_callback(verify:, authver_method:)
       @user = User.new(authver_method.profile_to_user_attrs(auth))
       @user.locale = selected_locale(omniauth_params) if selected_locale(omniauth_params)
 
-      SideFxUserService.new.before_create(@user, nil)
```
keep

kogre commented on back/app/controllers/web_api/v1/users_controller.rb:
```diff
@@ -99,8 +99,6 @@ def create
     @user.assign_attributes(permitted_attributes(@user))
     authorize @user
 
-    SideFxUserService.new.before_create(@user, current_user)
```
keep

kogre commented on back/app/services/invites_service.rb:
```diff
@@ -337,7 +337,6 @@ def validate_invite(invite)
   def save_invites invites
     ActiveRecord::Base.transaction do
       invites.each do |invite|
-        SideFxUserService.new.before_create(invite.invitee, invite.inviter)
```
keep

kogre commented on back/engines/commercial/admin_api/app/controllers/admin_api/users_controller.rb:
```diff
@@ -33,7 +33,7 @@ def show
     def create
       @user = User.new user_params
       @user.locale ||= AppConfiguration.instance.settings('core', 'locales').first
-      SideFxUserService.new.before_create @user, nil
```
keep

kogre commented on back/engines/free/user_confirmation/app/services/user_confirmation/patches/side_fx_user_service.rb:
```diff
@@ -1,12 +1,6 @@
 module UserConfirmation
   module Patches
     module SideFxUserService
-      def timestamp_registration(user);
-        return if AppConfiguration.instance.feature_activated?('user_confirmation')
-
-        super
-      end
-
```
Well spotted :+1:

kogre commented on back/spec/acceptance/invites_spec.rb:
```diff
@@ -282,7 +282,6 @@
         boulettos = json_response.dig(:included).select{|inc| inc[:id] == invite.invitee.id}&.first
         expect(boulettos&.dig(:attributes,:last_name)).to eq('Boulettos')
         expect(boulettos&.dig(:attributes,:invite_status)).to eq('accepted')
-        expect(invite.reload.invitee.registration_completed_at).to be_present  # when no custom fields
```
This is fine, but slightly better would be to change the test to make sure it's _not_ set instead.

jinjagit replied:
I have changed this to use `... .to eq(nil)` which seems clearer than using `... .not_to be_present`

kogre reviewed (changes requested):
Would keep the before_create hook, otherwise :ok_hand:

kogre commented:
Didn't see your questions yet, answering them after the review:

> Should we also remove the 'completed_registration' LogActivityJobs from here, here, and here? (since, I think, we now expect these conditions to never be true)...

You could indeed remove it from the invites side effect, and the create side effect, but shouldn't from the users update side effects.

> And should we add a 'completed_registration' LogActivityJob to the complete_registration action in users_controller.rb, perhaps in the if @user.save block? (I'm kinda surprised we don't do that already)
I think the `complete_registration` action calls the users side effects after_update, as it should, and that already takes care of it. Also the reason to not remove it from those side effects.

kogre commented:
> Is there any need for additional FE testing to cover the removals of BE responsibility for writing to registration_completed_at in certain special cases? Because I don't yet understand our FE testing, I can't answer this.

This is something we should maximally cover with e2e tests. I think we have quite some already, @luucvanderzee will know best. In any case, we should properly test, even if it's manually, with both FE and BE changes.

jinjagit committed:
Restore (empty version of) before_create hook and calls to it

@kogre explained: in this case, losing when exactly this side
effect should be called from the various places really is core
to the business logic (where in the system are we creating a user?)
that it could be painful to lose if we need new side effects on
before_create of a user.

jinjagit committed:
Include test expectation that registration_completed_at not set

jinjagit committed:
Remove 2 conditional 'completed_registration' LogActivityJob calls
